storage/sources/kafka: support custom partitioners (#28972)
MaterializeInc/materialize: doc/user/content/sql/create-sink/kafka.md

`TOPIC`                             | `text`              | The name of the Kafka topic to write to.
`COMPRESSION TYPE`                  | `text`              | The type of compression to apply to messages before they are sent to Kafka: `none`, `gzip`, `snappy`, `lz4`, or `zstd`.<br>Default: `none`.
`TRANSACTIONAL ID PREFIX`           | `text`              | The prefix of the transactional ID to use when producing to the Kafka topic.<br>Default: `materialize-{REGION ID}-{CONNECTION ID}-{SINK ID}`.

> `PARTITION HASHER`                  | expression          | A SQL expression returning a hash that can be used for partition assignment. See [Partitioning](#partitioning) for details.

Reviewer: How much do we need to bear the weight of choosing the hash function on the user? Could we expose this as `PARTITION KEY` with an additional `PARTITION HASH TYPE` that defaults to one of the hash functions?

Author: Yeah, it's a fair question. My philosophy was that we should build the most powerful and flexible thing for users first, and then if common patterns emerge, we can wrap them up in syntax sugar.

The reason I think it's useful to have full control over the partition strategy is so that you can exactly match the way your existing code partitions things. This is particularly relevant for composite keys. Suppose you want to hash both customer and order ID for the partition. How do you combine those two values together? Maybe your code was previously separating those two quantities with a `|` character:

```
murmur2(customer_id::text || '|' || order_id::text)
```

The design here lets you exactly match the way your code was already working.

If we were to offer a higher level interface like:

```
PARTITION KEY = (customer_id, order_id)
PARTITION HASH FUNCTION = murmur2
```

now there's a hard question to answer about how to combine those fields. Do we use the same format that you specified for the key? But hashing Avro/JSON values is fraught for all sorts of reason.

I do think we'll want to expose that higher level interface some day, but I don't think we'll have a clear answer to the question of how to hash composite partition keys until we see how the `PARTITION HASHER` is used in the wild.

Reviewer: Gotcha.

use. Materialize will ignore any headers in the map whose key starts with
`materialize-`.

> **Known limitations:**

Author: Leaving as is for now to align with the one-off `**Known limitation**` in the `PARTITION BY` section below, but ofc feel free to align on a different standard in the follow up PR it sounds like you have in the works.

For more details, see [the Kafka documentation](https://kafka.apache.org/documentation/).

### Partitioning

{{< warn-if-unreleased "v0.113" >}}

By default, Materialize assigns a partition to each message using the following

> scheme:

Reviewer: ```suggestion
strategy:
```

Author: Done.

  1. Encode the message's key in the specified format.

>   2. If the format uses a Confluent Schema Registry, strip out the
>      schema ID from the encoded bytes.

Reviewer: ```suggestion
  2. If the format uses a schema registry, strip out the schema ID from
    the encoded bytes.
```

Author: We consistently refer to the schema registry as a "Confluent Schema Registry" in the syntax and these docs—seems worth being consistent?

Reviewer: Seems to be a mix right now. In the `CREATE SOURCE` page, we use "schema registry" (which I prefer, because it generalizes...even though the syntax doesn't). Cool to leave it as is here, but we should decide on one term and stick with it across the board.

  3. Hash the remaining encoded bytes using [SeaHash].
  4. Divide the hash value by the topic's partition count and assign the
     remainder as the message's partition.

If a message has no key, all messages are sent to partition 0.

> You can configure a custom partitioning scheme with the `PARTITION HASHER`

Reviewer: ```suggestion
You can configure a custom partitioning strategy using the `PARTITION HASHER`
```

Author: Done.

For more details, see [the Kafka documentation](https://kafka.apache.org/documentation/).

### Partitioning

{{< private-preview />}}

By default, Materialize assigns a partition to each message using the following
strategy:

  1. Encode the message's key in the specified format.
  2. If the format uses a Confluent Schema Registry, strip out the
     schema ID from the encoded bytes.
  3. Hash the remaining encoded bytes using [SeaHash].
  4. Divide the hash value by the topic's partition count and assign the
     remainder as the message's partition.

If a message has no key, all messages are sent to partition 0.

You can configure a custom partitioning strategy with the `PARTITION BY` option,
which allows you to specify a SQL expression that computes a hash for each

> message that determines what partition to assign to the message, like so:

Author: Adopted the first part of this. I left the last bit in place, because I think "which determines what partition to assign to the message" reads more clearly (active vs passive) than "which determines the partition a message is assigned to."

```sql
-- General syntax.

> CREATE SINK ... INTO KAFKA CONNECTION <name> (PARTITION BY = <expression>) ...;

Author: Yeah, we don't have a good way of showing Kafka connection-specific options in the syntax diagram right now. :sadpanda:

See [Avro schema documentation](#avro-schema-documentation) for details
about the rules by which Materialize attaches `doc` fields to records.

#### Custom partitioning

Suppose your Materialize deployment stores data about customers and their
orders. You want to emit the order data to Kafka with upsert semantics so that
only the latest state of each order is retained. However, you want the data to
be partitioned by only customer ID (i.e., not order ID), so that all orders for
a given customer go to the same partition.

Create a sink using the `PARTITION BY` option to accomplish this:

```sql
CREATE SINK customer_orders
  FROM ...
  INTO KAFKA CONNECTION kafka_connection (
    TOPIC 'customer-orders',
    -- The partition hash includes only the customer ID, so the partition
    -- will be assigned only based on the customer ID.
    PARTITION BY = seahash(customer_id::text)
  )
  -- The key includes both the customer ID and order ID, so Kafka's compaction
  -- will keep only the latest message for each order ID.
  KEY (customer_id, order_id)
  FORMAT JSON
  ENVELOPE UPSERT;

> ```

Reviewer 2: this is a great example!
